utils/python/flash.py

```diff
@@ -0,0 +1,92 @@
+from subprocess import check_output
+import time
+
+def find_microbit():
+    """
+    Returns a path on the filesystem that represents the plugged in BBC
+    micro:bit that is to be flashed. If no micro:bit is found, it returns
+    None.
+    Works on Linux, OSX and Windows. Will raise a NotImplementedError
+    exception if run on any other operating system.
+    """
+    # Check what sort of operating system we're on.
+    if os.name == 'posix':
+        # 'posix' means we're on Linux or OSX (Mac).
+        # Call the unix "mount" command to list the mounted volumes.
+        mount_output = check_output('mount').splitlines()
+        mounted_volumes = [x.split()[2] for x in mount_output]
+        for volume in mounted_volumes:
+            if volume.endswith(b'MICROBIT'):
+                return volume.decode('utf-8')  # Return a string not bytes.
+    elif os.name == 'nt':
+        # 'nt' means we're on Windows.
+
+        def get_volume_name(disk_name):
+            """
+            Each disk or external device connected to windows has an attribute
+            called "volume name". This function returns the volume name for
+            the given disk/device.
+            Code from http://stackoverflow.com/a/12056414
+            """
+            vol_name_buf = ctypes.create_unicode_buffer(1024)
+            ctypes.windll.kernel32.GetVolumeInformationW(
+                ctypes.c_wchar_p(disk_name), vol_name_buf,
+                ctypes.sizeof(vol_name_buf), None, None, None, None, 0)
+            return vol_name_buf.value
+
+        #
+        # In certain circumstances, volumes are allocated to USB
+        # storage devices which cause a Windows popup to raise if their
+        # volume contains no media. Wrapping the check in SetErrorMode
+        # with SEM_FAILCRITICALERRORS (1) prevents this popup.
+        #
+        old_mode = ctypes.windll.kernel32.SetErrorMode(1)
+        try:
+            for disk in 'ABCDEFGHIJKLMNOPQRSTUVWXYZ':
+                path = '{}:\\'.format(disk)
+                #
+                # Don't bother looking if the drive isn't removable
+                #
+                if ctypes.windll.kernel32.GetDriveTypeW(path) != 2:
+                    continue
+                if os.path.exists(path) and \
+                        get_volume_name(path) == 'MICROBIT':
+                    return path
+        finally:
+            ctypes.windll.kernel32.SetErrorMode(old_mode)
+    else:
+        # No support for unknown operating systems.
+        raise NotImplementedError('OS "{}" not supported.'.format(os.name))
+
+def flash():
+    print("Flashing to micro:bit...")
+
+    mb_path = find_microbit()
+    hx_path = os.path.join(os.getcwd(), "MICROBIT.hex")
+
+    if mb_path:
+        new_hx_path = os.path.join(mb_path, "MICROBIT.hex")
+
+        print("micro:bit path: " + mb_path)
+        print("hex path: " + hx_path)
+
+        with open(hx_path, "r") as hex_read:
+            user_hex = hex_read.read()
```

Comment:
What happens if the file doesn't exist? If there is an exception bubbling up it might be better to try/catch it and provide a user friendly error message.

Reply:
d17ccea